vllm/compilation/fusion.py

```diff
@@ -0,0 +1,219 @@
+                             epsilon=1e-5)
+
+    # result
+    return at[1]
+
+
+def rms_pattern_residual_static(result: torch.Tensor, input: torch.Tensor,
+                                residual: torch.Tensor, weight: torch.Tensor,
+                                scale: torch.Tensor):
+    at = auto_functionalized(torch.ops._C.fused_add_rms_norm.default,
+                             input=input,
+                             residual=residual,
+                             weight=weight,
+                             epsilon=1e-5)
+    at1 = auto_functionalized(torch.ops._C.static_scaled_fp8_quant.default,
+                              result=result,
+                              input=at[1],
+                              scale=scale)
+
+    # result, residual
+    return at1[1], at[2]
+
+
+def rms_replacement_residual_static(result: torch.Tensor, input: torch.Tensor,
+                                    residual: torch.Tensor,
+                                    weight: torch.Tensor, scale: torch.Tensor):
+    at = auto_functionalized(
+        torch.ops._C.fused_add_rms_norm_static_fp8_quant.default,
+        result=result,
+        input=input,
+        residual=residual,
+        weight=weight,
+        scale=scale,
+        epsilon=1e-5)
+    # result, residual
+    return at[1], at[2]
+
+
+my_patterns = PatternMatcherPass()
+
+
+def empty_bf16(*args, **kwargs):
+    return torch.empty(*args, **kwargs, dtype=torch.bfloat16, device="cuda")
+
+
+def empty_fp8(*args, **kwargs):
+    return torch.empty(*args,
+                       **kwargs,
+                       dtype=torch.float8_e4m3fn,
+                       device="cuda")
+
+
+def get_patterns():
+    my_patterns = PatternMatcherPass(pass_name="fusion_pass")
+
+    inputs = [
+        empty_fp8(5, 4),
+        empty_bf16(5, 4),
+        empty_bf16(5, 4),
+        empty_bf16(1, 5),
+        torch.empty(1, 1, device="cuda", dtype=torch.float32)
+    ]
+    register_replacement(rms_pattern_static, rms_replacement_static, inputs,
+                         fwd_only, my_patterns)
+
+    matches = []
+
+    def record_match_fn(match: Match):
+        matches.append(match)
+        return False
+
+    # with residual
+    inputs = [
+        empty_fp8(5, 4),
+        empty_bf16(5, 4),
+        empty_bf16(5, 4),
+        empty_bf16(1, 5),
+        torch.empty(1, 1, device="cuda", dtype=torch.float32)
+    ]
+    register_replacement(rms_pattern_residual_static,
+                         rms_replacement_residual_static,
+                         inputs,
+                         fwd_only,
+                         my_patterns,
+                         extra_check=record_match_fn)
+
+    return my_patterns, matches
+
+
+def process_matches(matches, graph: torch.fx.Graph):
+    for match in matches:
+        nodes = list(graph.nodes)
+        last_node_in_match = max(match.nodes, key=lambda x: nodes.index(x))
+        with graph.inserting_after(last_node_in_match):
```

Comment:
This section feels like it needs a comment. Something to describe why we are adding these getitem nodes along with the general tuple structure of the auto_functionalized output.